File: lib/bin/main/dbassistant/layers/PromptPreparationLayer.kt
package dbassistant.layers

import dbassistant.interfaces.ILayer
import dbassistant.layers.LayerCargo

class PromptPreparationLayer(
    private val promptSize: Int
) : ILayer{
    
    override fun pass(cargo: LayerCargo): LayerCargo {
        val input = cargo.take() as String
        val specialCharacter = Regex("[^A-z0-9 @]")

        var spaced = ""
        for(char in input.replace("\"","")){
            if(specialCharacter.matches(char.toString())){
                spaced += " $char "
            }
            else{
                spaced += char
            }
        }
        spaced = spaced.replace(Regex(" +")," ")
        val symbols = spaced.split(" ").toMutableList()
        symbols.add("[END]")
        
        cargo.put(symbols.toList())
        return cargo
    }

}

File: assistant/layers/PromptPreparationLayer.ts
import ILayer from '../interfaces/ILayer';
import Tokenizer from '../tokens/Tokenizer';


/**
 * @input string
 * @output string[]
 */
export default class PromptPreparationLayer implements ILayer {
    private promptSize: number;

    constructor(promptSize: number) {
        this.promptSize = promptSize;
    }
    
    process(inputString: string): string[] {
        const specialCharacter = /[^A-z0-9 _@]/;
        let spaced = "";
        for (const char of inputString) {
            if (specialCharacter.test(char)) {
                spaced += " " + char + " ";
            } else {
                spaced += char;
            }
        }
        spaced = spaced.replace(/ +/g, ' ');
        const symbols = spaced.split(" ").filter(symbol => symbol !== "");

        symbols.push("[END]");
        return symbols;
    }
}

